sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulAggregate.scala

```diff
@@ -82,40 +82,62 @@ case class StateStoreRestoreExec(
 case class StateStoreSaveExec(
     keyExpressions: Seq[Attribute],
     stateId: Option[OperatorStateId],
+    returnAllStates: Option[Boolean],
```

Comment:
Similar to stateId, it is injected later after creation of the physical Spark plan, and before execution. See IncrementalExecution, and how stateId is injected using the prepare rule. 

I agree that this structure is something we can improve, but that would require quite a bit of refactoring and better done later.